### PR TITLE
Support weight for extension

### PR DESCRIPTION
**Pull request by pradpnayak**
Overview
----------------------------------------
Support weight for extensions

### CODE REVIEW

**Comment by mlutfy** [comment]
Would this help? https://docs.civicrm.org/dev/en/latest/hooks/#symfony-method

**Comment by colemanw** [comment]
I agree with @mlutfy the Drupal-style hooks and module weights are flawed because giving each module a weight will affect *all* hooks implemented by that module.
The symfony-style listeners are better because you can set the priority for each one.

**Comment by totten** [comment]
Also relevant : https://docs.civicrm.org/dev/en/latest/hooks/usage/symfony/#priorities

File: CRM/Extension/Mapper.php [hunk 1]
```diff
@@ -304,6 +304,7 @@ public function getActiveModuleFiles($fresh = FALSE) {
         FROM civicrm_extension
         WHERE is_active = 1
         AND type = "module"
+        ORDER BY weight ASC
```
Comment (totten) [inline CRM/Extension/Mapper.php:307]: I agree it's usually preferrable to set per-listener Symfony `$priority` (ie you don't get boxed-in when diff listeners need different priorities).

But... I do think the PR is right to hint that something is off wrt the `civicrm_extension` ordering. The problem is that the default order is *undefined*... so the de-facto order depends on internal+arbitrary DBMS details. You can get two different sites with the same extensions - but they fire in different orders. (*IIRC, it's common for happenstance to loosely mean "first-installed to last-installed", which isn't crazy - but it's not guaranteed, and any specific ordering is hard to reproduce.*)

From that POV, adding a `weight` with default `weight=0` still leaves the ordering undefined. (There's no way to differentiate a bunch of `weight=0`s.) Alternatively, sorting on `full_name` or `name` would be simple + reproducible.

```sql
ORDER BY full_name
```

But I imagine that ordering looks rather different from the  *de-facto* orderings.

Sorting on the dependency-graph might be a little smoother for an install-base where the de-facto orderings most closely approximate "first-installed to last-installed". Something like this, perhaps:

```
cv ev '$m=CRM_Extension_System::singleton()->getManager(); $k=array_keys($m->getStatuses()); sort($k); return array_flip($m->findInstallRequirements($k));'
```

(*Of course, ou'd want to store that ordering somewhere... like in `civicrm_extension.weight` or maybe just a cache...*)

**Comment by pradpnayak** [comment]
With the Symfony-style listeners i need to add priority for each hook, which i was trying to avoid. It really works if one of the hook should be called first and other last but not incase most of the hook should be executed last unless i add listeners for each of them every time.

**Comment by mlutfy** [comment]
> With the Symfony-style listeners i need to add priority for each hook, which i was trying to avoid. It really works if one of the hook should be called first and other last but not incase most of the hook should be executed last unless i add listeners for each of them every time.

I'm curious, could you add more information about the use-case?

If we add a specific behaviour, then we need to make sure it's well documented, has tests, etc.

**Comment by colemanw** [comment]
This PR has no description so we're all just guessing what it aims to do and why.

I think the right procedure is to close this and ask @pradpnayak to open a GL issue for discussion. That issue needs a clear problem statement. If that gets `concept-approved` then this or a similar PR can be reopened.

**Pull request closed without merging**